File: dsa-problems/Sorting/solutions/javascript/SRT-002-kth-missing-positive-blocks.js
class Solution {
  solve(arr, queries) {
    const results = [];
    const n = arr.length;
    
    for (const [k, b] of queries) {
      // Use BigInt for safety as k*b can exceed 2^53
      const m = BigInt(k) * BigInt(b);
      
      let low = 0;
      let high = n - 1;
      let idx = -1;
      
      while (low <= high) {
        const mid = Math.floor((low + high) / 2);
        const missingCount = BigInt(arr[mid]) - BigInt(mid + 1);
        if (missingCount < m) {
          idx = mid;
          low = mid + 1;
        } else {
          high = mid - 1;
        }
      }
      
      // Result is m + idx + 1
      // idx is number, m is BigInt. 
      // idx + 1 is the count of numbers in arr that are <= result
      results.push(Number(m + BigInt(idx + 1)));
    }
    return results;
  }
}

const fs = require("fs");

const input = fs.readFileSync(0, "utf8").trim();
if (!input) process.exit(0);
const data = input.split(/\s+/);
let idx = 0;
const n = parseInt(data[idx++], 10);
const q = parseInt(data[idx++], 10);
const arr = [];
for (let i = 0; i < n; i++) {
  arr.push(parseInt(data[idx++], 10));
}
const queries = [];
for (let i = 0; i < q; i++) {
  const k = data[idx++];
  const b = data[idx++];
  queries.push([k, b]);
}
const solution = new Solution();
const results = solution.solve(arr, queries);
console.log(results.join("\n"));

File: dsa-problems/Sorting/solutions/cpp/SRT-002-kth-missing-positive-blocks.cpp
#include <vector>
#include <algorithm>
#include <iostream>

using namespace std;

class Solution {
public:
    vector<long long> solve(const vector<int>& arr, const vector<pair<long long,long long>>& queries) {
        vector<long long> results;
        results.reserve(queries.size());
        int n = arr.size();
        
        for (const auto& q : queries) {
            long long m = q.first * q.second;
            
            int low = 0, high = n - 1;
            int idx = -1;
            
            while (low <= high) {
                int mid = low + (high - low) / 2;
                long long missingCount = arr[mid] - (mid + 1);
                if (missingCount < m) {
                    idx = mid;
                    low = mid + 1;
                } else {
                    high = mid - 1;
                }
            }
            
            results.push_back(m + idx + 1);
        }
        return results;
    }
};

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    int n, q;
    if (!(cin >> n >> q)) return 0;
    vector<int> arr(n);
    for (int i = 0; i < n; i++) {
        cin >> arr[i];
    }
    vector<pair<long long, long long>> queries;
    queries.reserve(q);
    for (int i = 0; i < q; i++) {
        long long k, b;
        cin >> k >> b;
        queries.push_back({k, b});
    }
    Solution solution;
    vector<long long> results = solution.solve(arr, queries);
    for (long long v : results) {
        cout << v << "\n";
    }
    return 0;
}
